### PR TITLE
ci: Upload wheels to PyPI using GitHub Actions

### LINKED ISSUE #10: ci cleanup
Some suggested cleanup to build-wheels:

* Currently there are three test jobs and three wheel build jobs; now that there's a ci.yml, the test jobs are either no longer needed, or should be moved.
* The cibuildwheel and twine (should use the pypa publish action) should be moved out of a custom script and the action form should be used (can be pinned and updated with Dependabot, too)
* macos-13 (last Intel) and macOS-14 (first AS) images should be used.
* The custom arch stuff in pyproject.toml can then be removed, as native will work fine.

Also, `"wheel"` should not be listed in pyproject.toml, and `backend = setuptools.build_meta` is missing under `[build-system]`.

See https://learn.scientific-python.org/development/guides/repo-review/?repo=thaler-lab%2FWasserstein&branch=master for other ideas.

### PR DESCRIPTION
**Pull request by j-s-ashley**
Addresses part of #10 

Resolves #10 

```
* Use trusted publishers to publish to PyPI on GitHub release via 
  https://github.com/pypa/gh-action-pypi-publish.
   - Require publishing to be in the GitHub deployment environment 'pypi'.

Co-authored-by: Matthew Feickert <matthew.feickert@cern.ch>
```

### CODE REVIEW

File: .github/workflows/build-wheels.yml
Comment (matthewfeickert) [inline .github/workflows/build-wheels.yml:68]: We'll need someone who has owner control over the PypI page (@rikab) to enable trusted publishers for this to work. Instructions here: https://docs.pypi.org/trusted-publishers/adding-a-publisher/

**Review by matthewfeickert — CHANGES_REQUESTED** [review]
This is good, but there should also be some conditional guards on it like in https://github.com/scikit-hep/boost-histogram/blob/8692d97e8134d1553196f6192b2de438d88a5203/.github/workflows/wheels.yml#L133 so that it doesn't attempt to upload to PyPI for every merge to `master`.

**Comment by j-s-ashley** [comment]
Per #7:
> Have a seperate publishing workflow that downloads the artifacts and then publishes them using [PyPI's Trusted Publishers](https://learn.scientific-python.org/development/guides/gha-wheels/#publishing).

I'll kick this change from `Builds` to its own workflow.

File: .github/workflows/publish.yml
Comment (matthewfeickert) [inline .github/workflows/publish.yml:25]:
```suggestion
        uses: pypa/gh-action-pypi-publish@release/v1
        with:
          print-hash: true
```

**Review by matthewfeickert — COMMENTED** [review]
This just adds a check sum, and you're already getting the triggers. So between the two of these this should be good.

**Review by matthewfeickert — COMMENTED** [review]
So following the conversation with Henry today the issues that I was concerned about with regards to downloading the correct packages were realized, so that we basically need to just append this to the https://github.com/thaler-lab/Wasserstein/blob/444905b415e84de30d903f949b21e464919ecd3d/.github/workflows/build-wheels.yml workflow.

**Comment by j-s-ashley** [comment]
@henryiii @matthewfeickert This is ready for review

File: .github/workflows/build-wheels.yml
Comment (matthewfeickert) [inline .github/workflows/build-wheels.yml:92]:
```suggestion
        uses: pypa/gh-action-pypi-publish@release/v1
        with:
          print-hash: true
          attestations: true
```

This just adds some additional options from https://github.com/pypa/gh-action-pypi-publish.

**Review by matthewfeickert — COMMENTED** [review]
I think this is good now and I've added some suggested additional options. Probably good to get a quick review from @henryiii even if it is after merge.

File: .github/workflows/build-wheels.yml
Comment (henryiii) [reply .github/workflows/build-wheels.yml:92]: You could also add artifact generation. You'd also need `attestations: write` permission for that. But that could all be in a followup?

**Review by henryiii — APPROVED** [review]

File: .github/workflows/build-wheels.yml
Comment (matthewfeickert) [inline .github/workflows/build-wheels.yml:78]:
```suggestion
    # Restrict to the environment set for the trusted publisher
    environment:
```

**Pull request merged**
